app/agents/voice/automatic/tools/juspay/analytics.py

```
@@ -387,6 +389,10 @@ async def create_euler_offer(params: FunctionCallParams):
             await params.result_callback({"error": "Authentication token is missing. Cannot create offer."})
             return
 
+        if not is_user_authorized(email):
```

For Juspay tools, it should come from Juspay ACL.
Create an user with only read access for Offers and try creating an offer.
That should throw an Error

---

@swaroopvarma1 It is an immediate fix to block unauthorized write actions for the users specified in the env. Ideally, this enforcement should come directly from Juspay ACL, but to prevent further ops loss, we need to put this safeguard in place right away. 

cc: @murdore

---

Then we should write logic to skip `write tools initialization` for that user session. 
More tools are being added, adding this check to all the tools is unnecessary.

---

Keep an env ACTIONS_BLACKLISTED_USERS